DOCSP-9565 crud upsert (#71)
mongodb/docs-java: source/fundamentals/crud/write-operations/upsert.txt

.. default-domain:: mongodb

Overview
--------

You can insert or update documents by passing a query filter to the 
:java-docs:`updateOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`, 
:java-docs:`updateMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`
or :java-docs:`replaceOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument,com.mongodb.client.model.ReplaceOptions)>` methods

Each method returns an ``AcknowledgedUpdateResult`` that tells us:

- The number of documents that matched the query filter
- The number of documents that got modified 
- The ``_id`` of the document that got upserted 

These methods update documents that match your query filter, but
inserts a document if the there are no matches when you set the upsert

> to true.  

Reviewer: This still feels off. You state that documents can be inserrted with update or replace, but don't state how until after what the methods return. Additionally, is the information on what the methods return necessary here? Placing it so high implies it is important, but it doesn't seem that important.

Is this about what the methods return or about upsert behavior?

> In the case there are multiple matches in your query filter for
> ``updateOne()`` or ``replaceOne()``, the method will update the first
> occurrence of a match in the collection.

Reviewer: ```suggestion
When multiple documents match the query filter supplied to
``updateOne()`` or ``replaceOne()``, the first document that matches
will be modified.
```

suggestion: simplify

> Set an upsert:

Reviewer: I don't think "set" is the correct word here. It's an option that is specified.

~~~~~~~~~~~~~~

Use :java-docs:`UpdateOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateOptions.html>`

> as a third parameter to the ``updateOne()`` and ``updateMany()``

Reviewer: ```suggestion
as the third parameter to the ``updateOne()`` and ``updateMany()``
```
suggestion: Did you mean "the"? Are there other things that can be specified as a 3rd parameter, or are the options the only thing that can be specified as the 3rd parameter to these methods? Same throughout.

Author: For each method, there are various forms of them.  Some variations are:
- ClientSession clientSession, Bson filter, List<? extends Bson> update
- ClientSession clientSession, Bson filter, Bson update, UpdateOptions updateOptions
- Bson filter, Bson update
- Bson filter, Bson update, UpdateOptions updateOptions

colors of paint. The store had their annual online sale resulting in the
following documents in their ``paint_inventory`` collection: 

.. code-block:: json

    { "_id": { "$oid": "606b4cfbcd83be7518b958da" }, "color": "red", "qty": 5 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958db" }, "color": "purple", "qty": 8 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958dc" }, "color": "blue", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958dd" }, "color": "white", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958de" }, "color": "yellow", "qty": 6 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958df" }, "color": "pink", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958e0" }, "color": "green", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958e1" }, "color": "black", "qty": 8 }

The store is restocking their items.  The first shipment that arrived
are ten cans of orange paint and we need to update this in the collection.

To update orange, query the ``paint_inventory`` collection where the
``color`` is ``"orange"``, set the update for ``qty`` to ``10``, set
``UpdateOptions.upsert()`` to ``true`` and pass the query to the
``updateOne()`` method: 

> 

Reviewer: Looking at this again, I noticed there's some information here that doesn't really add anything. To summarize:

1. The store received a shipment of inventory.
2. The store uses upsert operations so that new inventory is added, or existing inventory is updated in a single operation.

If that's the case, maybe something along the lines of
The store received a fresh shipment and needs to update their inventory. By using an upsert, existing documents will be updated and new documents will be created in a single operation, simplifying the logic in their inventory control application.

The first item in the shipment is ten cans of orange paint. To update the inventory, or insert, query the ``paint_inventory`` collection for the appropriate color, specify an ``increment`` operation of 10 for the ``qty`` field in the update, and specify ``true`` to ``UpdateOptions.upsert()``.

Something along the lines of that. I'm unsure the value of stating the ``updateOne`` method since the example shows the method used.

.. default-domain:: mongodb

Overview
--------

> You often use insert and update operations for applications that store
> and modify data. Sometimes, you need to choose between an insert or
> update depending on whether the document exists. 
> 
> To insert or update documents, pass a query filter to the 
> :java-docs:`updateOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`, 
> :java-docs:`updateMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`
> or :java-docs:`replaceOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument,com.mongodb.client.model.ReplaceOptions)>`
> methods. 
> 
> These methods update documents that match your query filter,
> but inserts a document if the there are no matches when you set the
> upsert to true. When multiple documents match the query filter supplied to
> ``updateOne()`` or ``replaceOne()``, the first document that matches
> gets updated.

Reviewer 2: **issue:** This does not get to the point of _this_ page. In reading this, you should move this content to a separate page for **Update** operations, then get into Upsert.

To introduce Upsert in the introduction, I would recommend something closer to this:
```suggestion
Applications don't only store data. Data changes over time. You can use Update operations
to modify existing data. Update operations assume the data that you want to change exists.
This creates a wrinkle when developing your application. You would write logic to check for
the data first. MongoDB offers an easier option called **Upsert**.
```

> Specify an upsert:
> ~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Specify an Upsert:
~~~~~~~~~~~~~~~~~~
```

> Use :java-docs:`UpdateOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateOptions.html>`
> as the third parameter to the ``updateOne()`` and ``updateMany()``
> methods.
> 
> Use :java-docs:`ReplaceOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/ReplaceOptions.html>`
> as the third parameter to the ``replaceOne()`` method.

Reviewer 2: **suggestion:** Reword to make this more concise.

```suggestion
- Add the following third parameter to the ``updateOne()`` or ``updateMany()``
  methods: :java-docs:`UpdateOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/UpdateOptions.html>`.

- Add the following third parameter to the ``replaceOne()``
  method: :java-docs:`ReplaceOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/ReplaceOptions.html>`.
```

Example
-------

> The following example is about a paint store that sells eight different

Reviewer 2: **suggestion:** Remove passive voice.

```suggestion
In the following example, a paint store sells eight different
```

colors of paint. The store had their annual online sale resulting in the

> following documents in their ``paint_inventory`` collection: 

Reviewer 2: **question:** Are these totals before or after the sale? It would help to clarify that.

**suggestion:** Refactor to separate the "human" language from the "programming" language. That is, state the situation with the store. Then say something like, `The store's database now shows the following...`

.. default-domain:: mongodb

Overview
--------

You can update documents by passing a query filter to the 
:java-docs:`updateOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,java.util.List)>`, 
:java-docs:`updateMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,java.util.List)>`
or :java-docs:`replaceOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument)>` methods

Each method returns an ``AcknowledgedUpdateResult`` that tells us:

> - the number of documents that matched the query filter
> - the number of documents that got modified 
> - the id of the document that got upserted 

Reviewer: ```suggestion
- The number of documents that matched the query filter
- The number of documents that got modified 
- The ``_id`` of the document that got upserted 
```

> An upsert is an update to a document or insert of a document.  An update

Reviewer: nitpick:

This is a rough transition. It might be better to add an introduction to the concept before stating what an upsert is. Right now it reads "these methods accept a filter, and return some Thing. An upsert is another thing..."

occurs when there is at least one match to your query filter.  An insert
occurs when there is no match to your query filter.

> To set an upsert:

Reviewer: set, or specify? Also, the heading should probably be: Verb indefinite article Noun

~~~~~~~

The store is restocking their items.  The first shipment that arrived
are ten cans of orange paint and we need to update this in the collection.

To update orange, query the ``paint_inventory`` collection where the
``color`` is ``"orange"``, set the update for ``qty`` to ``10``, set
``UpdateOptions.upsert()`` to ``true`` and pass the query to the
``updateOne()`` method: 

.. literalinclude:: /includes/fundamentals/code-snippets/InsertUpdate.java
   :language: java
   :dedent:
   :start-after: begin updateOneExample
   :end-before: end updateOneExample

The method returns:

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=0, modifiedCount=0, upsertedId=BsonObjectId{ value=606782721601f9443bc1c7a7 }} 

The following shows the documents in the ``paint_inventory`` collection:

.. code-block:: json   

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 3, "color": "blue", "qty": 0 }
    { "_id": 4, "color": "white", "qty": 0 }
    { "_id": 5, "color": "yellow", "qty": 6 }
    { "_id": 6, "color": "pink", "qty": 0 }
    { "_id": 7, "color": "green", "qty": 0 }
    { "_id": 8, "color": "black", "qty": 8 }

>     { "_id": { "$oid": "606782721601f9443bc1c7a7" }, "color": "orange", "qty": 10 }

Reviewer: issue: use consistent _id values (oid). It is a serious anti-pattern to mix their types.

``replaceOne()`` method:

.. literalinclude:: /includes/fundamentals/code-snippets/InsertUpdate.java
   :language: java
   :dedent:
   :start-after: begin updateManyExample
   :end-before: end updateManyExample

The method returns:

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=1, modifiedCount=1, upsertedId=null } 

The following shows the documents in the ``paint_inventory`` collection:

.. code-block:: json   

    { "_id": 1, "color": "red", "qty": 25 }
    { "_id": 2, "color": "purple", "qty": 28 }
    { "_id": 3, "color": "blue", "qty": 20 }
    { "_id": 4, "color": "white", "qty": 20 }
    { "_id": 5, "color": "yellow", "qty": 26 }
    { "_id": 6, "color": "pink", "qty": 20 }
    { "_id": 7, "color": "green", "qty": 20 }
    { "_id": 8, "color": "black", "qty": 28 }
    { "_id": { "$oid": "60678b161601f9443bc31df6" }, "color": "orange", "qty": 10 }
    
See the following documentation for more information about the updating
a document:

- :manual:`db.collection.updateOne() </reference/method/db.collection.updateOne/>`
- :manual:`db.collection.updateMany() </reference/method/db.collection.updateMany/>`

> - :manual:`db.collection.replaceOne() </reference/method/db.collection.replaceOne/>`

Reviewer: question: I'm unsure what this content has to do with upserting? Might be better to remove it to focus on the main topic.

.. default-domain:: mongodb

Overview
--------

Insert or update documents by passing a query filter to the 
:java-docs:`updateOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`, 
:java-docs:`updateMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson,com.mongodb.client.model.UpdateOptions)>`
or :java-docs:`replaceOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument,com.mongodb.client.model.ReplaceOptions)>`
methods. These methods update documents that match your query filter,
but inserts a document if the there are no matches when you set the
upsert to true. When multiple documents match the query filter supplied to
``updateOne()`` or ``replaceOne()``, the first document that matches

> gets updated.

Reviewer: 🤔 

This part still feels off to me. If someone didn't know about upsert behavior, the introduction sentence might leave them feeling like they missed something earlier.

Maybe a gentler introduction that states documents can be inserted using update or replace using upsert behavior, and then how to specify upsert behavior, and then why it might be useful?

Use :java-docs:`ReplaceOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/ReplaceOptions.html>`
as the third parameter to the ``replaceOne()`` method.

Example
-------

The following example is about a paint store that sells eight different
colors of paint. The store had their annual online sale resulting in the
following documents in their ``paint_inventory`` collection: 

.. code-block:: json

    { "_id": { "$oid": "606b4cfbcd83be7518b958da" }, "color": "red", "qty": 5 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958db" }, "color": "purple", "qty": 8 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958dc" }, "color": "blue", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958dd" }, "color": "white", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958de" }, "color": "yellow", "qty": 6 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958df" }, "color": "pink", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958e0" }, "color": "green", "qty": 0 }
    { "_id": { "$oid": "606b4cfbcd83be7518b958e1" }, "color": "black", "qty": 8 }

The store received a fresh shipment and needs to update their inventory.
The first item in the shipment is ten cans of orange paint.

> To update or insert orange, query the ``paint_inventory`` collection

Reviewer: ```suggestion
To update the inventory, query the ``paint_inventory`` collection
```
suggestion: simplify

It is obvious the subject is orange here because of the instructions that follow and the example.

where the ``color`` is ``"orange"``, specify an update to ``increment`` the
``qty`` field by ``10``, and specify ``true`` to
``UpdateOptions.upsert()``: 

.. literalinclude:: /includes/fundamentals/code-snippets/InsertUpdate.java
   :language: java
   :dedent:
   :start-after: begin updateOneExample
   :end-before: end updateOneExample

The method returns:

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=0, modifiedCount=0, upsertedId=BsonObjectId{ value=606b4cfc1601f9443b5d6978 }} 

This ``AcknowledgedUpdateResult`` tells us:

- Zero documents matched our query filter
- Zero documents in our collection got modified 

> - A document with an ``_id`` of  ``606b4cfc1601f9443b5d6978`` got upserted

Reviewer: Nice way to show this is useful
